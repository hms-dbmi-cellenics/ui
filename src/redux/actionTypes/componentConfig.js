/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const RESET_CONFIG = `${COMPONENT_CONFIG}/reset`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;
const PLOT_DATA_ERROR = `${COMPONENT_CONFIG}/plotDataError`;

export {
  LOAD_CONFIG, UPDATE_CONFIG, RESET_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR,
};
