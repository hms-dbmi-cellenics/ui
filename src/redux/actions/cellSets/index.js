import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
