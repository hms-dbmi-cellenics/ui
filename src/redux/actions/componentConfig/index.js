import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import fetchPlotDataWork from 'redux/actions/componentConfig/fetchPlotDataWork';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchPlotDataWork,
  updatePlotData,
  resetPlotConfig,
};
