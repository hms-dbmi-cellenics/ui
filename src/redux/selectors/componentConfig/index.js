import getPlotConfigs from './getPlotConfigs';
import getSelectedMetadataTracks from './getSelectedMetadataTracks';
import getGroupSlidesBy from './getGroupSlidesBy';

/* eslint-disable-next-line import/prefer-default-export */
export { getPlotConfigs, getSelectedMetadataTracks, getGroupSlidesBy };
