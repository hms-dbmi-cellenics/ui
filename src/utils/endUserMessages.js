export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of processing the experiment\'s processing run.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: 'Created a new cluster. Check the `Data Management` window.',
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  NO_CLUSTERS: 'No custom cell sets available, to create more go to data-exploration.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
};
