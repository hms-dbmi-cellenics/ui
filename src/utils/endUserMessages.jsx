export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  ERROR_CREATING_SAMPLE: 'We couldn\'t create your sample. Please try uploading it again',
  ERROR_CREATING_PROJECT: 'We couldn\'t create your project.',
  ERROR_LOADING_PROJECT: 'We couldn\'t load your project.',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data.',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_NORMALIZED_EXPRESSION_MATRIX: 'We couldn\'t download the normalized expression matrix.',
  ERROR_NO_MATCHING_CELLS_NORMALIZED_EXPRESSION_MATRIX: 'No cells match requested filters.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_DATA_PROCESSING_SETTINGS: 'We couldn\'t get the data processing settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of the experiment\'s data processing.',
  ERROR_FETCH_BACKGROUND_GENE_EXP: 'We couldn\'t get the background gene expression',
  ERROR_FETCH_MARKER_GENES: 'We couldn\'t get the marker genes',
  ERROR_FILE_CATEGORY: 'The selected file name does not match the expected category.',
  ERROR_FILES_FOLDER: 'Only files contained in a folder are accepted.',
  ERROR_METADATA_MULTIPLE_FILES: 'You need to upload a single file in .tsv format',
  ERROR_METADATA_WRONG_FORMAT: 'The file does not seem to be in the correct tsv key-value format.',
  ERROR_CELL_LEVEL_COLUMN: 'The .tsv file needs to contain the column "barcode"',
  ERROR_OBJ2S_MULTIPLE_FILES: 'Only a single file is accepted for uploads.',
  ERROR_OBJ2S_EXISTING_FILE: 'Please remove existing file before uploading a new one.',
  ERROR_OBJ2S_MAX_FILE_SIZE: 'file size must be less than 15 GB.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_NO_PERMISSIONS: 'You do not have permissions to perform that action.',
  ERROR_LAUNCH_PATHWAY: 'We couldn\'t launch pathway analysis',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_BEGIN_SAMPLE_FILE_UPLOAD: 'We couldn\'t begin the files upload',
  ERROR_UPDATE_SERVER_ON_UPLOAD_STATE: 'We couldn\'t update the upload state in our servers. Try refreshing the page.',
  ERROR_SAMPLE_TECHNOLOGY: 'Selected technology does not match across all samples. Select the right technology and try again.',
  ERROR_SAVING_PLOT_CONFIG: 'We couldn\'t save your plot configuration.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  ERROR_DOWNLOADING_OBJ2S: 'We couldn\'t get the processed object.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: (
    <>
      Created a new cluster. Check &nbsp;
      <b>Custom cell sets</b>
      &nbsp; in the &nbsp;
      <b>Cell sets and Metadata </b>
      tile.
    </>),
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  SUCCESS_CELL_SETS_ANNOTATED: 'Cell sets have been annotated. Your cell sets have changed.',
  WORK_REQUEST_TIMED_OUT_RETRY: 'The task is taking longer than expected, please retry.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
  SHARE_SUCESS_NOTIFICATION_FAILURE: 'The project was shared, but we werenâ€™t able to notify the new collaborator',
  SHARE_FAILURE: 'We weren\'t able to share the project',
  ERROR_CLONING_DEFAULT: 'We couldn\'t clone the project.',
  ERROR_CLONING_PIPELINE_LOCKED: 'We couldn\'t clone the project because data processing is currently running, wait until it finishes and try again',
  ERROR_CELL_SETS_ANNOTATION_FAILED: 'ScType annotation unsuccessful. Please ensure dataset contains gene symbols.',
};
