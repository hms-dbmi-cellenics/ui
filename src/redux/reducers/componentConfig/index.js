import {
  UPDATE_CONFIG, CONFIGS_REPLACED,
  SAVE_CONFIG, LOAD_CONFIG, RESET_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR, TRAJECTORY_NODES_SELECTION_UPDATED,
} from 'redux/actionTypes/componentConfig';

import { CELL_CLASS_DELETE } from 'redux/actionTypes/cellSets';

import initialState from 'redux/reducers/componentConfig/initialState';
import loadConfig from 'redux/reducers/componentConfig/loadConfig';
import updateConfig from 'redux/reducers/componentConfig/updateConfig';
import configsReplaced from 'redux/reducers/componentConfig/configsReplaced';
import saveConfig from 'redux/reducers/componentConfig/saveConfig';
import resetConfig from 'redux/reducers/componentConfig/resetConfig';
import plotDataLoaded from 'redux/reducers/componentConfig/plotDataLoaded';
import plotDataLoading from 'redux/reducers/componentConfig/plotDataLoading';
import plotDataError from 'redux/reducers/componentConfig/plotDataError';
import trajectoryNodesUpdated from 'redux/reducers/componentConfig/trajectoryNodesUpdated';

import cellClassDelete from 'redux/reducers/componentConfig/cellClassDelete';

const plotsReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOAD_CONFIG:
      return loadConfig(state, action);
    case UPDATE_CONFIG:
      return updateConfig(state, action);
    case CONFIGS_REPLACED:
      return configsReplaced(state, action);
    case RESET_CONFIG:
      return resetConfig(state, action);
    case SAVE_CONFIG:
      return saveConfig(state, action);
    case PLOT_DATA_LOADED:
      return plotDataLoaded(state, action);
    case PLOT_DATA_LOADING:
      return plotDataLoading(state, action);
    case PLOT_DATA_ERROR:
      return plotDataError(state, action);
    case TRAJECTORY_NODES_SELECTION_UPDATED:
      return trajectoryNodesUpdated(state, action);
    case CELL_CLASS_DELETE: {
      return cellClassDelete(state, action);
    }
    default:
      return state;
  }
};

export default plotsReducer;
