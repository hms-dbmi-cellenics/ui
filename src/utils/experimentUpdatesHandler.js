import updateCellSetsClustering from 'redux/actions/cellSets/updateCellSetsClustering';
import { updateProcessingSettingsFromQC, loadedProcessingConfig } from 'redux/actions/experimentSettings';
import { updateBackendStatus } from 'redux/actions/backendStatus';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import pushNotificationMessage from 'utils/pushNotificationMessage';

import { loadCellSets } from 'redux/actions/cellSets';
import endUserMessages from './endUserMessages';

const updateTypes = {
  QC: 'qc',
  GEM2S: 'gem2s',
  WORK_RESPONSE: 'WorkResponse',
};

const experimentUpdatesHandler = (dispatch) => (experimentId, update) => {
  if (update.status) {
    dispatch(updateBackendStatus(experimentId, update.status));
  }

  if (update.response?.error) {
    return;
  }

  switch (update.type) {
    case updateTypes.QC: {
      return onQCUpdate(update, dispatch, experimentId);
    }
    case updateTypes.GEM2S: {
      return onGEM2SUpdate(update, dispatch, experimentId);
    }
    case updateTypes.WORK_RESPONSE: {
      return onWorkResponseUpdate(update, dispatch, experimentId);
    }
    default: {
      console.log(`Error, unrecognized message type ${update.type}`);
    }
  }
};

const onQCUpdate = (update, dispatch, experimentId) => {
  const { input, output } = update;

  const processingConfigUpdate = output.config;

  if (processingConfigUpdate) {
    dispatch(updateProcessingSettingsFromQC(
      input.taskName,
      processingConfigUpdate,
      input.sampleUuid,
      false,
    ));

    Object.entries(output.plotData).forEach(([plotUuid, plotData]) => {
      dispatch(updatePlotData(plotUuid, plotData));
    });
  }

  // If the pipeline finished we have a new clustering, so fetch it
  if (update.status.pipeline.status === 'SUCCEEDED') {
    dispatch(loadCellSets(experimentId, true));
  }
};

const onGEM2SUpdate = (update, dispatch, experimentId) => {
  const processingConfig = update?.item?.processingConfig;
  if (processingConfig) {
    dispatch(loadedProcessingConfig(experimentId, processingConfig, true));
  }
};

const onWorkResponseUpdate = (update, dispatch, experimentId) => {
  const { request: { body: { name: workRequestName } }, response: { error } } = update;

  if (error) throw new Error(error);

  if (workRequestName === 'ClusterCells') {
    dispatch(updateCellSetsClustering(experimentId));
    pushNotificationMessage('success', endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED);
  }

  if (workRequestName === 'GetExpressionCellSets') {
    dispatch(loadCellSets(experimentId, true));
    pushNotificationMessage('success', endUserMessages.SUCCESS_NEW_CLUSTER_CREATED);
  }
};

export default experimentUpdatesHandler;
