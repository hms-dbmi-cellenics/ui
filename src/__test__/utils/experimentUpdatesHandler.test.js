import experimentUpdatesHandler, { updateTypes } from 'utils/experimentUpdatesHandler';
import fake from '__test__/test-utils/constants';
import pushNotificationMessage from 'utils/pushNotificationMessage';

import { updateCellSetsClustering, loadCellSets } from 'redux/actions/cellSets';
import { updateProcessingSettingsFromQC, loadedProcessingConfig, updatePipelineVersion } from 'redux/actions/experimentSettings';
import { updateBackendStatus } from 'redux/actions/backendStatus';
import { updatePlotData, replaceLoadedConfigs } from 'redux/actions/componentConfig';
import endUserMessages from 'utils/endUserMessages';

jest.mock('redux/actions/cellSets/updateCellSetsClustering');
jest.mock('redux/actions/experimentSettings', () => ({
  updateProcessingSettingsFromQC: jest.fn(),
  loadedProcessingConfig: jest.fn(),
  updatePipelineVersion: jest.fn(),
}));
jest.mock('redux/actions/backendStatus/updateBackendStatus');
jest.mock('redux/actions/componentConfig/updatePlotData');
jest.mock('redux/actions/cellSets/loadCellSets');
jest.mock('redux/actions/componentConfig/replaceLoadedConfigs');

jest.mock('utils/pushNotificationMessage');
jest.spyOn(global.console, 'error');

const mockDispatch = jest.fn();

const triggerExperimentUpdate = (update) => {
  const actionCreator = experimentUpdatesHandler(mockDispatch);
  actionCreator(fake.EXPERIMENT_ID, update);
};

const mockError = {
  error: true,
  errorCode: 'MOCK_ERROR_CODE',
  userMessage: 'mockUserMessage',
};

const mockBackendStatusError = {
  pipeline: {
    status: 'ERROR',
  },
};

const mockQcUpdate = {
  type: updateTypes.QC,
  input: {
    taskName: 'mockTaskName',
    sampleUuid: 'mockSampleUuid',
  },
  output: {
    config: {
      processingConfigUpdate: {
        mockProcessingConfigUpdate: 'mockProcessingConfigUpdate',
      },
    },
    plotData: {
      mockPlotUuid: 'mockPlotData',
    },
  },
  response: {
    error: false,
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
  pipelineVersion: 2,
};

const mockGem2sUpdate = {
  type: updateTypes.GEM2S,
  item: {
    processingConfig: {
      mockProcessingConfig: 'mockProcessingConfig',
    },
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
};

const mockWorkResponseUpdate = {
  type: updateTypes.WORK_RESPONSE,
  request: {
    body: {
      name: 'WorkResponse',
    },
  },
  response: {
    error: false,
  },
};

const mockObj2sUpdate = {
  type: updateTypes.OBJ2S,
  item: {
    processingConfig: {
      mockProcessingConfig: 'mockProcessingConfig',
    },
  },
  status: {
    seurat: {
      status: 'RUNNING',
    },
  },
};

describe('ExperimentUpdatesHandler', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Triggers properly for GEM2S updates ', () => {
    const mockUpdate = mockGem2sUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - Loaded processingConfig
    const updateParams = loadedProcessingConfig.mock.calls[1];
    expect(updateParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(2);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for OBJ2S updates ', () => {
    const mockUpdate = mockObj2sUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for OBJ2S updates ', () => {
    const mockUpdate = mockObj2sUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  it('Loads cell sets if Seurat pipeline completes ', () => {
    const mockUpdate = {
      ...mockObj2sUpdate,
      status: {
        seurat: {
          status: 'SUCCEEDED',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - load cellsets on seurat finish
    expect(loadCellSets).toHaveBeenCalledTimes(1);
    const loadCellSetsParams = loadCellSets.mock.calls[0];
    expect(loadCellSetsParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(2);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for QC updates ', () => {
    const mockUpdate = mockQcUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];

    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - update processin settings
    expect(updateProcessingSettingsFromQC).toHaveBeenCalledTimes(1);
    const updateParams = updateProcessingSettingsFromQC.mock.calls[0];
    expect(updateParams).toMatchSnapshot();

    // Dispatch 3 - update pipeline version
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion.mock.calls).toMatchSnapshot();

    // Dispatch 4 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(4);

    // Does not load cell sets
    expect(loadCellSets).not.toHaveBeenCalled();

    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Handles qc errors correctly with config update', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      output: {
        ...mockQcUpdate.output,
        plotData: undefined,
      },
      response: mockError,
      status: mockBackendStatusError,
    };

    triggerExperimentUpdate(mockUpdate);

    // Performs all the updates it can
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    expect(updateProcessingSettingsFromQC).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);

    // Doesn't try to update the plot data
    expect(updatePlotData).not.toHaveBeenCalled();

    expect(global.console.error).toHaveBeenCalledTimes(1);
    expect(global.console.error).toMatchSnapshot();
  });

  it('Handles qc errors correctly without config update', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      output: undefined,
      response: mockError,
      status: mockBackendStatusError,
    };

    triggerExperimentUpdate(mockUpdate);

    // Performs all the updates it can
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);

    // Doesn't try to update the plot data or config
    expect(updateProcessingSettingsFromQC).not.toHaveBeenCalled();
    expect(updatePlotData).not.toHaveBeenCalled();

    expect(global.console.error).toHaveBeenCalledTimes(1);
    expect(global.console.error).toMatchSnapshot();
  });

  it('Loads cell sets if QC pipeline completes ', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      status: {
        pipeline: {
          status: 'SUCCEEDED',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    // Dispatch 3 - update pipeline version
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion.mock.calls).toMatchSnapshot();

    // Dispatch 3 - load cellsets on pipeline finish
    expect(loadCellSets).toHaveBeenCalledTimes(1);
    const loadCellSetsParams = loadCellSets.mock.calls[0];
    expect(loadCellSetsParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(5);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for WorkResponse - ClusterCells updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'ClusterCells',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update cell set clustering
    expect(updateCellSetsClustering).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED,
    );
  });

  it('Triggers properly for WorkResponse - GetExpressionCellSets updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'GetExpressionCellSets',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Load cell sets
    expect(loadCellSets).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_NEW_CLUSTER_CREATED,
    );
  });

  it('Triggers properly for PlotConfigRefresh updates', () => {
    const normalizedMatrixConfig = {
      id: 'normalized-matrix',
      updatedConfig: {
        sample: [],
        louvain: [],
        metadata: [],
        scratchpad: [],
      },
    };

    const mockUpdate = {
      type: updateTypes.PLOT_CONFIG_REFRESH,
      updatedConfigs: [normalizedMatrixConfig],
    };

    triggerExperimentUpdate(mockUpdate);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(replaceLoadedConfigs).toHaveBeenCalledTimes(1);

    expect(replaceLoadedConfigs).toHaveBeenCalledWith([normalizedMatrixConfig]);
  });
});
