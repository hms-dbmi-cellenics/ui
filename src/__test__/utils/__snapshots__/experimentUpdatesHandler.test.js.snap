// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentUpdatesHandler Handles qc errors correctly with config update 1`] = `
[MockFunction] {
  "calls": [
    [
      "Experiment updates error:",
      {
        "input": {
          "sampleUuid": "mockSampleUuid",
          "taskName": "mockTaskName",
        },
        "output": {
          "config": {
            "processingConfigUpdate": {
              "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
            },
          },
          "plotData": undefined,
        },
        "pipelineVersion": 2,
        "response": {
          "error": true,
          "errorCode": "MOCK_ERROR_CODE",
          "userMessage": "mockUserMessage",
        },
        "status": {
          "pipeline": {
            "status": "ERROR",
          },
        },
        "type": "qc",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ExperimentUpdatesHandler Handles qc errors correctly without config update 1`] = `
[MockFunction] {
  "calls": [
    [
      "Experiment updates error:",
      {
        "input": {
          "sampleUuid": "mockSampleUuid",
          "taskName": "mockTaskName",
        },
        "output": undefined,
        "pipelineVersion": 2,
        "response": {
          "error": true,
          "errorCode": "MOCK_ERROR_CODE",
          "userMessage": "mockUserMessage",
        },
        "status": {
          "pipeline": {
            "status": "ERROR",
          },
        },
        "type": "qc",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  1`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "pipeline": {
      "status": "SUCCEEDED",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  2`] = `
[
  "mockPlotUuid",
  "mockPlotData",
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  3`] = `
[
  [
    "testae48e318dab9a1bd0bexperiment",
    2,
  ],
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  4`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  true,
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if Obj2s pipeline completes  1`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "obj2s": {
      "status": "SUCCEEDED",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if Obj2s pipeline completes  2`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  true,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for GEM2S updates  1`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "pipeline": {
      "status": "RUNNING",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for GEM2S updates  2`] = `undefined`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  1`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "pipeline": {
      "status": "RUNNING",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  2`] = `
[
  "mockTaskName",
  {
    "processingConfigUpdate": {
      "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
    },
  },
  "mockSampleUuid",
  false,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  3`] = `
[
  [
    "testae48e318dab9a1bd0bexperiment",
    2,
  ],
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  4`] = `
[
  "mockPlotUuid",
  "mockPlotData",
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for OBJ2S updates  1`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "obj2s": {
      "status": "RUNNING",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for OBJ2S updates  2`] = `
[
  "testae48e318dab9a1bd0bexperiment",
  {
    "obj2s": {
      "status": "RUNNING",
    },
  },
]
`;
